### PR TITLE
Improve abc module and builtin function decorators (without ParamSpec)

### PR DESCRIPTION
**Pull request by srittau**
* Make classmethod and staticmethod generic.
* abstractclassmethod and abstractstaticmethod are classes, not
  functions.
* Remove mypy-specific comments.

### CODE REVIEW

**Comment by srittau** [comment]
Simplified version of #5682 that should work with mypy. #5682 will then only contain the problematic changes.

**Comment by srittau** [comment]
@rchen152 Do you know what the problem here is? Is `classmethod` special-cased by pytype? Weirdly, staticmethod does seem to work.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -109,19 +111,19 @@ class object:
...
-class staticmethod(object):  # Special, only valid as a decorator.
-    __func__: Callable[..., Any]
+class staticmethod(Generic[_FuncT]):  # Special, only valid as a decorator.
+    __func__: _FuncT
     __isabstractmethod__: bool
-    def __init__(self, f: Callable[..., Any]) -> None: ...
+    def __init__(self, __f: _FuncT) -> None: ...
     def __new__(cls: Type[_T], *args: Any, **kwargs: Any) -> _T: ...
-    def __get__(self, obj: _T, type: Optional[Type[_T]] = ...) -> Callable[..., Any]: ...
+    def __get__(self, obj: _T, type: Type[_T] | None = ...) -> _FuncT: ...
```
Comment (Akuli) [inline stdlib/builtins.pyi:119]: Static methods can be looked up as `TheClass.method`, and in that case there is no associated instance:

```suggestion
    def __get__(self, obj: _T | None, type: Type[_T] | None = ...) -> _FuncT: ...
```

**Comment by rchen152** [comment]
@srittau We probably just need to copy the changes to builtins.pyi to pytype/stubs/. I'll give that a try.

**Comment by srittau** [comment]
@rchen152 Please don't change anything permanent for now, though. We are still figuring out whether this PR is the best course of action.

**Comment by rchen152** [comment]
I finally got around to testing out making classmethod and staticmethod generic in pytype (https://github.com/google/pytype/commit/6ca00db931300abfe0804a777094d93a8833eab2). And yep, once I do that, pytype_test passes on this PR. So if you do decide to go with this, the pytype_test fix should be pretty straightforward.

**Comment by srittau** [comment]
Looking at the primer output:

* `pydantic` just needs updated annotations (another case for python/typing#307)
* `graphql-core` is a corner case: assigning the result `staticmethod()` to a mutable variable and the argument function having a default value. Can easily be solved with an explicit annotation.
* The `pydantic` errors are regrettable, but only affects the deprecated `@abstractclassmethod` and `@abstractstaticmethod` decorators.

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -73,7 +74,7 @@ _T4 = TypeVar("_T4")
 _T5 = TypeVar("_T5")
 _TT = TypeVar("_TT", bound="type")
 _TBE = TypeVar("_TBE", bound="BaseException")
-_R = TypeVar("_R")  # Return-type TypeVar
+_FuncT = TypeVar("_FuncT", bound=Callable[..., Any])
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:77]: Where is this used?

Comment (srittau) [reply stdlib/builtins.pyi:77]: Good catch, it's a left-over from before the merge.

**Comment by srittau** [comment]
I am marking it as draft. If this gets accepted, we need to get in the pytype fix first.

**Comment by AlexWaygood** [comment]
Making the TypeVar for `classmethod`/ `staticmethod` covariant is definitely worth doing.

When I looked at the `abc` decorators in #6552, I concluded the primer diff was problematic enough to not make those changes worth it. They look like false positives to me — WDYT?

**Comment by srittau** [comment]
They look like false positives indeed, but I'm not sure why they are flagged. I wouldn't be surprised if `@classmethod` is special-cased by mypy, but `@abstractclassmethod` is not. But since `@abstractclassmethod` should be split anyway, and this is an easy fix, I would prefer the correct annotation over the incorrect one.

**Comment by AlexWaygood** [comment]
Yeah you're right — they're deprecated anyway, so the obvious solution if you start getting false positives from mypy is just to stop using them!

**Comment by AlexWaygood** [comment]
We could probably now add `ParamSpec` to `staticmethod` without any difficulty? (Not sure that we can do `classmethod` without support for `Concatenate`.)

**Comment by srittau** [comment]
Marking as ready to review again. @rchen152: Could you make the necessary changes to pytype?

**Comment by rchen152** [comment]
> Marking as ready to review again. @rchen152: Could you make the necessary changes to pytype?

Will do! Hopefully I'll be able to make this change and release it this week; either way, I'll update in a couple days.

**Comment by rchen152** [comment]
Alright, pytype-2022.1.7 contains the necessary changes for pytype_test to play nicely with this PR.

**Comment by srittau** [comment]
Looks good now.

File: stdlib/abc.pyi [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import sys
 from _typeshed import SupportsWrite
-from typing import Any, Callable, TypeVar
+from collections.abc import Callable
+from typing import Any, Generic, Type, TypeVar
```
Comment (JelleZijlstra) [inline stdlib/abc.pyi:4]:
```suggestion
from typing import Any, Generic, TypeVar
```

File: stdlib/builtins.pyi [hunk 4]
```diff
@@ -112,26 +112,26 @@ class object:
     def __dir__(self) -> Iterable[str]: ...
     def __init_subclass__(cls) -> None: ...
 
-class staticmethod(Generic[_R]):
-    __func__: Callable[..., _R]
+class staticmethod(Generic[_R_co]):  # Special, only valid as a decorator.
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:115]:
```suggestion
class staticmethod(Generic[_R_co]):
```

File: stdlib/builtins.pyi [hunk 5]
```diff
@@ -112,26 +112,26 @@ class object:
...
         __name__: str
         __qualname__: str
-        __wrapped__: Callable[..., _R]
-        def __call__(self, *args: Any, **kwargs: Any) -> _R: ...
+        __wrapped__: Callable[..., _R_co]
+        def __call__(self, *args: Any, **kwargs: Any) -> _R_co: ...
 
-class classmethod(Generic[_R]):
-    __func__: Callable[..., _R]
+class classmethod(Generic[_R_co]):  # Special, only valid as a decorator.
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:126]:
```suggestion
class classmethod(Generic[_R_co]):
```

**Review by AlexWaygood — COMMENTED** [review]
Looks like these comments were accidentally added back in the merge as well

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Pull request merged**